### ISSUE TITLE
Add authorization (permissions) support

### ISSUE THREAD

**Issue by alexjreich** [nxus/users#20]
Should use the current `role` field on the User model to define access controls.

Need a `authorizeRoute` API for defining which roles have access to a resource.

```
import {users} from 'nxus-users'

users.authorizeRoute('/myRoute', 'adminRole') // All users AdminRole and above have access
```

Need a `authorize` EJS helper for filtering front-end components based on user role:

```
<% authorize('adminRole', () => { %>
<button>My Admin Button</button>
<% }) %>
```

**Comment by loppear** [comment]
Suggestion from a couple fronts is to support more specific permissions/rights:
- Route handlers would declare specific permissions, which are separately associated with roles
- Lets us build screens for managing roles or user-specific permissions rather than hard-coded
- UI would also need to check for specific permissions
- Additionally, we need some version of object-level permissions ("User has permission to view objects, and specifically THIS object")
  - Currently suggesting we punt on storing per-object ACLs, though I could be tallked out of it

## Approaches to registering
1. Separate route registration
2. Wrap route handler
3. Wrap route registration

Separate registration for permission middleware. Have to repeat route -  other approaches makes e.g. ViewController harder to extend with perms externally though (a case where repeating the route is the handle you have).

```
    router.route('get', this.base+'/add-users/:id', ::this._addUser)
    userPermissions.allow(this.base+'/add-users/:id', "initiatives-manage-users")
```

Wrapper for router handler. Would need support in router for promised-handlers. Assumes permissions don't need to be middleware, which is probably true (with specific permissions, less likely to want to protect a whole route sub-path at once.)

```
    router.route('get', this.base,
                 userPermissions.allow("initiatives-view", ::this._list))
```

 Convenience wrapper for router.route - challenge is managing combined args in this case. My least favorite.

```
    userPermissions.route('initiatives-view-comments',
                          'get', this.base+'/comments/:id', ::this._comments)
```

## Additional API thoughts

 Other args we might want to pass when registering permissions:
- a default role that gets this perm (this might be misleading in use as data is changed, but bootstraps role data.)
- an object check function (the light-weight answer to per-object permissions

```
    userPermissions.allow('initiatives-manage-users',
                          "Initiative Manager",
                          (user, obj) => {return obj.managers.includes(user)})
```

The object check probably requires cooperation in the handler method unless we also have the middleware/wrapper know how to query the object.

```
    _addUser (req, res) {
      this.models.Initiative.findOne(req.params.id).then((initiative) => {
        // contributed by userPermissions, 
        req.checkObjectPermission(initiative)
          .then(...)
          .catch(PermissionDenied, (e) => { res.status(403)...})
      })
    }
```

**Comment by alexjreich** [comment]
I like option 2, because I think we’ll also use that type of wrapper function as an EJS helper for displaying different UI elements.  That gives it some consistency.

``` ejs
<h1>Test</h1>
<% permissions.allow('initiatives-manage-users', () => { %>
  <button>My Button</button>
<% }) %>
```

I think we’ll need option 1 as well for the role specific authoritzation (the admin route for example will need to be accessible to only a specific set of users).

**Comment by loppear** [comment]
Forgot to mention, it's important that the style be able to be registered during startup so that we can be aware of all the permissions - no checks ONLY within a route handler or EJS, for instance.

I also like 2 style-wise the most, but supporting both 1 and 2 probably means two separate implementations (2 can't be internally implemented by middleware cause it doesn't know the route).

**Comment by alexjreich** [comment]
A note on API naming.  I would propose we have one `permissions` module that then exposes user/team permissions.

`permissions.allowUser(...)`
`permissions.allowTeam(...)`

**Comment by alexjreich** [comment]
@loppear what does 

> it's important that the style be able to be registered during startup so that we can be aware of all the permissions

mean exactly?

If its required, then I would recommend we add an explicit registration method and/or throw an exception if you try to reference a permission that doesn't exist.

``` javascript
permissions.role('my-role')
permissions.permission('my-role', 'my-permission')
```
1. With the caveat that these can be added/updated in run time, and we'll need an admin UI for managing the permissions and assignment to roles.
   1. Roles/Permissions will be stored in a database, so we'll need the setup of roles/permissions to be possible after `app.startup`.

**Comment by loppear** [comment]
Yes, "aware of all permissions at startup" just means that in order to have a UI for managing permission-role mappings we need explicit registration. Permissions themselves are not dynamic/data, just the mappings: code should define all permissions the application uses, and needs to do so somewhere in constructor/startup not just a code path checking for permission somewhere in a route or template.

As noted in conversation, allowTeam probably isn't needed, just a mechanism for assigning permissions/roles to teams and calculating the full set of a user's permissions (for a user-specific check).

**Comment by loppear** [comment]
So, updated API proposal:

```
import {permissions} from 'nxus-users' // submodule exposed here? 

// permissions.allow(permissionName, [handler], [objectCheckFunction], [defaultRole])

// In a module constructor defining routes
router.route("/where", permissions.allow("permission-name", ::this.somewhere, (obj, user) => {...}, "Manager"))

// Same, no object check or default role
router.route("/there", permissions.allow("there-name", ::this.there))

// In a module that only wants to define a permission for some template bit but not route handler
// Just defines permission
permissions.allow("permission-name")
```

Do we have a sane case for a third-party module to register a permission for a module's route? (The middleware case.) I think so, for ViewController if it doesn't become permissions-aware (but it should) and for e.g. view user profile - `nxus-users` may define a permission for this, but an app could decide to put an additional permission on it for requiring that you be a member of the user's team...

```
// In my app module, adding an additive permission to a route (prefix)
permissions.allowRoute('team-profile-view', '/users/profile/view', (obj, user) => {...})
```

For EJS helper, I'm inclined to just make this an accessor on req.user, thoughts? It needs to be sync, but it probably makes sense for a permissions middleware to pre-load a user's permissions on each authenticated request anyways. But this could alternatively be available at `permissions.allow` though I'd still make it sync not callback?

```
 <% if (req.user.allowed("permission-name")) { %>
...
 <% if (req.user.allowed("object-permission", obj)) { %>
```

**Comment by alexjreich** [comment]
👍let's give it a go

**Comment by loppear** [comment]
Righto, the route handler wrapping case would need to be:

```
  permissions.allow("permission-name", ::this.somewhere).then((handler) => {
    router.route("/where", handler)
  })
// or partial application
  permissions.allow("permission-name", ::this.somewhere).then(_.partial(router.route, "/where"))
```

So we'll see whether we actually end up liking the middleware one more.

**Comment by ScottMaxson** [comment]
@loppear: for what it's worth, I've not found object level permissions to be that useful. But maybe I just haven't run into a compelling use case yet. I see this will accommodate it, or not, which is great.

**Comment by loppear** [comment]
@ScottMaxson my view on object permissions, embodied in this approach: I too have found them to rarely be convenient or needed on a system-wide basis - flexibility of arbitrary per-object permissions are really hard to reason about for devs and for users, and are usually only there to provide a granular system for implementing some broader relational rules that have all the update-problem you'd expect. So those rules are where some support for object-level/object-specific statements is helpful, expressed as "has this permission and this _relationship_ to the object in question". A compromise, just lifting the common workaround in a role/permission-only system (adding checks to route handlers in addition to the permissions system) into the API of the permission system here so they can be expressed together.

**Comment by loppear** [comment]
Walking back my previous comment, this compromise of `objectCheck(obj, user)` is silly. The relationship I really want to have on (some key) objects is `(user, role, object)` - specific in each case. Where I've seen object level permissions get wonky is in needing to restate and maintain that relationship for child objects (e.g. all the comments on a post, depending on whether the user is the author of a post).

But definitely going to need to work in a way for (user, role) to be associated with other models and checked.

# Updated object-level proposal

Scratch the `checkObject` function.
`nxus-users` provides a base class for the three-way intersection table `(user, role, object)`:

```
import {ObjectRoleModel} from 'nxus-users'

export default ObjectRoleModel.extend({
  identity: 'thing-user-role',
  attributes: {
    object: {
      model: 'thing',
      via: 'roles'
    }
  }
})
```

Registering permissions that belong at an object level now takes a mapping of url params to the identity of this intersection collection:

```
permissions.allow('things-manage', '/path/to/:thing', {'thing': 'thing-user-role'}, 'Thing Manager')
```

This allows the permission middleware / handler-wrapper to query that collection based on the route and retrieve roles to add to the user's overall roles.

And granting object permissions looks like:

```
permissions.getRoles().then((roles) => {
       thing.roles.add({user: req.user, role: roles['Thing Manager'], object: thing})
}
```

# TODO
1. In-template checks for permissions against a specific object will need to include that object's populated user-roles records (TBD). Easy for further checks against the route's object, but...
2. The objectParams mapping argument doesn't address how to get from a child entity route (thing.comments) to the permissions on Thing for "Manage Comments", say. Probably needs to accept an promised function that will return roles for that sort of manual avoidance. Only an issue if you don't have the central object's id in the route.
